0010-R-components-healthcheck.md

```diff
@@ -0,0 +1,51 @@
+# Components Healthcheck In Dapr
+
+* Author(s): Deepanshu Agarwal (@DeepanshuA)
+* State: Draft
+* Updated: 07/20/2023
+
+## Overview
+As a user, I should be able to enquire the health of dapr system, which is possible currently - but this health check should also include health of connected dapr components as well.
+
+## Tenets
+1. It should be extensible i.e. tomorrow if any feature to check health of Actors etc. is required, it should not require a new endpoint.
+
+## Current Scenario
+There are many components in Dapr which don't yet implement Ping. 
+Ping is not mandatory to be implemented by Components, which is the correct behavior, as it could lead to false positives.
+For this components health-check, components not implementing Ping will be omitted out.
+
+## API Design
+### Endpoint:
+Instead of an additional endpoint, the Approach underneath works with a query parameter, in addition to `healthz` endpoint.
+
+Following are the possible responses for `healthz` API:
+
+| HTTP | Response Codes | 
+| -------- | -------- | 
+| 204     | dapr is healthy     | 
+| 500     | dapr is not healthy     | 
+
+Hence, if dapr is healthy, then the response code changes, as per the enlisted cases below.
+If dapr is NOT healthy, then the compoenents Health check should anyways NOT be checked.
+
+http://localhost:3500/v1.0/healthz?include_components=true
+
+### Approach: 
+- Maintain a cache with status of all components loaded successfully and keep updating this cache in a background go routine at a configurable `pingUpdateFrequency`. By default, `pingUpdateFrequency` to be 5 minutes.
```

Comment:
Some thoughts:

- 5 mins seems a lot? That would mean the time to detect a failure can be as high as 5 mins. Maybe that's just my opinion.
- Ping updates may be more frequent if the component is un-healthy, since we want to see if we can recover faster

Reply:
- I agree that 5 minutes seems higher. I have updated that to 30 seconds.
- I like this point but do we need another configuration for this? One more config will be too much, so rather gave logic that in case of unhealthy, it will Ping every `pingUpdateFrequency` / 3.

Reply:
If un-healthy we should use an exponential backoff (with a limit of `pingUpdateFrequency`) IMHO, to avoid overloading a service.